pkg/kubectl/cmd/util/helpers.go

```diff
@@ -400,22 +404,70 @@ func DumpReaderToFile(reader io.Reader, filename string) error {
 	return nil
 }
 
+// UpdateApplyAnnotation serializes the VersionedObject of an info, puts it into
+// an annotation on the Object that can be used by RunApply, and returns both the
+// original and modified contents of the annotation.
+func UpdateApplyAnnotation(info *resource.Info) ([]byte, []byte, error) {
+	// Get the current annotations from the object.
+	annotations, err := info.Mapping.MetadataAccessor.Annotations(info.Object)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// If there weren't any, then make a new map of annotations.
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	// Retrieve the original configuration from the annotations, if present.
+	original := []byte(annotations[ApplyAnnotation])
+
+	var modified []byte
+	if info.VersionedObject != nil {
```

Comment:
Good question, and one that I thought about and addressed. The solution here is subtle, and probably deserves a comment. The VersionedObject is the input to the command, not the object from the server. It therefore doesn't contain the annotation unless the user has put it there. I thought about stripping it out if they did, but that seems contrary to the philosophy of letting the user define the desired behavior. Most users will never think to add the annotation on their own. Those who do will do so consciously, and will observe that it gets nested. However, unless they add it back to their input the next time around, it will never nest more than one deep.

So, bottom line... no, it doesn't recurse or even embed the previous annotation by default, but the user can achieve one or multiple levels of embedded annotations if they explicitly set it up that way.